File: vns-python/pyvns/decompiler.py
from typing import Any

from .content import ContentManager
from .naming import Naming


# 反编译器
class Decompiler:
    # tag formate
    TAG_STARTS: str = "["
    TAG_ENDS: str = "]"

    # 如果输入字符串为None，则将其转换为null
    @staticmethod
    def __to_str_in_case_null(text: str | None) -> str:
        return text if text is not None else "null"

    @classmethod
    def decompile(cls, _data: dict[str, dict[str, dict[str, Any]]], out: str) -> None:
        # 如果数据为空，则直接返回
        if _data is None or len(_data) <= 0:
            return
        # 初始化视觉小说数据管理模块
        _content: ContentManager = ContentManager()
        # 把数据更新到管理模块中
        _content.update(_data)
        # 用于储存结果的列表
        _results: list[str] = [
            "# Fundamental parameters\n",
            f"{cls.TAG_STARTS}id{cls.TAG_ENDS}{_data['id']}\n",
            f"{cls.TAG_STARTS}language{cls.TAG_ENDS}{_data['language']}\n",
        ]
        # 历遍所有section
        for _section in _data:
            # 更新视觉小说数据管理模块的当前位置
            _content.set_id("head")
            _content.set_section(_section)
            # 写入当前部分的名称
            _results.append(f"\n{cls.TAG_STARTS}section{cls.TAG_ENDS}{_section}\n")

            while True:
                _current_dialog: dict[str, Any] = _content.get_dialog()
                # 处理注释
                comments: list[str] = _current_dialog.pop("comments", [])
                if len(comments) > 0:
                    _results.append("\n")
                    _results.extend(f"// {_note}\n" for _note in comments)
                # 写入讲话人名称
                _results.append(
                    "null:\n"
                    if len(_content.current.narrator) == 0
                    else _content.current.narrator + ":\n"
                )
                # 写入对话
                for _sentence in _content.current.contents:
                    _results.append("- " + _sentence + "\n")

                """如果下列内容有变化，则写入"""
                # 写入背景
                if (
                    _content.previous is None
                    or _content.current.background_image
                    != _content.previous.background_image
                ):
                    if (
                        _content.previous is None
                        or _content.previous.next.type != "scene"
                    ):
                        _results.append(
                            f"{cls.TAG_STARTS}bgi{cls.TAG_ENDS}{cls.__to_str_in_case_null(_content.current.background_image)}\n"
                        )
                    else:
                        _results.append(
                            f"{cls.TAG_STARTS}scene{cls.TAG_ENDS}{cls.__to_str_in_case_null(_content.current.background_image)}\n"
                        )
                # 写入背景音乐
                if (
                    _content.previous is None
                    or _content.previous.background_music
                    != _content.current.background_music
                ):
                    _results.append(
                        f"{cls.TAG_STARTS}bgm{cls.TAG_ENDS}{cls.__to_str_in_case_null(_content.current.background_music)}\n"
                    )
                # 写入当前立绘
                if (
                    _content.previous is None
                    or _content.previous.character_images
                    != _content.current.character_images
                ):
                    if len(_content.current.character_images) == 0:
                        _results.append(f"{cls.TAG_STARTS}hide{cls.TAG_ENDS}*\n")
                    elif (
                        _content.previous is None
                        or len(_content.previous.character_images) == 0
                    ):
                        _line: str = f"{cls.TAG_STARTS}display{cls.TAG_ENDS}"
                        for _characterName in _content.current.character_images:
                            _line += Naming(_characterName).name + " "
                        _results.append(_line.rstrip() + "\n")

                if _content.current.has_next():
                    if isinstance(_content.current.next.target, str):
                        _content.set_id(_content.current.next.target)
                    elif isinstance(_content.current.next.target, list):
                        # branch is currently not supported
                        pass
                else:
                    break

        # 写入停止符
        _results.append(f"\n{cls.TAG_STARTS}end{cls.TAG_ENDS}\n")

        # 保存反编译好的脚本
        with open(out, "w+", encoding="utf-8") as f:
            f.writelines(_results)

File: vns-cpp/decompiler.cpp
#include <fstream>
#include <vector>
#include "naming.hpp"
#include "decompiler.hpp"


// If input string is nullptr, convert it to "null"
std::string Decompiler::to_str_in_case_null(const std::string &text)
{
    return text.empty() ? "null" : text;
}

void Decompiler::decompile(const std::unordered_map<std::string, std::any> &data, const std::filesystem::path &out_path)
{
    // If data is empty, return directly
    if (data.empty())
        return;

    // Initialize visual novel data management module
    DialoguesManager content_manager;
    // Update data to the management module
    content_manager.set_data(std::any_cast<DialogueSectionsDataType>(data.at("dialogs")));

    // string stream to store results
    std::stringstream result_ss;

    // add comment for Fundamental parameters
    result_ss << "# Fundamental parameters\n";
    // create id line
    result_ss << TAG_STARTS << "id" << TAG_ENDS << to_str_in_case_null(std::any_cast<std::string>(data.at("id")))
              << "\n";
    // create language line
    result_ss << TAG_STARTS << "language" << TAG_ENDS
              << to_str_in_case_null(std::any_cast<std::string>(data.at("language"))) << "\n";

    // Iterate through all sections
    for (const std::string &k: data | std::views::keys)
    {
        // Update the current position of the visual novel data management module
        content_manager.set_id("head");
        content_manager.set_section(k);

        // create section line
        result_ss << "\n" << TAG_STARTS << "section" << TAG_ENDS << k << "\n";

        while (true)
        {
            const DialogueDataType &current_dialog = content_manager.get_current_dialogue();
            // Process comments
            if (const auto &comments = std::get<std::vector<std::string>>(
                        current_dialog.at("comments")); !comments.empty())
            {
                result_ss << "\n";
                for (const std::string &note: comments)
                    result_ss << "// " << note << "\n";
            }
            // Write speaker's name
            result_ss << (content_manager.get_current()->narrator.empty() ? "null"
                                                                          : content_manager.get_current()->narrator)
                      << ":\n";
            // Write dialogue
            for (const auto &sentence: std::get<std::vector<std::string>>(current_dialog.at("contents")))
                result_ss << "- " << sentence << "\n";

            // If the following content has changed, write it
            // Write background
            if (content_manager.get_previous() == nullptr ||
                content_manager.get_current()->background_image != content_manager.get_previous()->background_image)
            {
                if (content_manager.get_previous() == nullptr ||
                    content_manager.get_previous()->next.get_type() != "scene")
                    result_ss << TAG_STARTS << "bgi" << TAG_ENDS
                              << to_str_in_case_null(content_manager.get_current()->background_image) << "\n";
                else
                    result_ss << TAG_STARTS << "scene" << TAG_ENDS
                              << to_str_in_case_null(content_manager.get_current()->background_image) << "\n";
            }
            // Write background music
            if (content_manager.get_previous() == nullptr ||
                content_manager.get_current()->background_music != content_manager.get_previous()->background_music)
            {
                result_ss << TAG_STARTS << "bgm" << TAG_ENDS
                          << to_str_in_case_null(content_manager.get_current()->background_music) << "\n";
            }
            // Write current character images
            if (content_manager.get_previous() == nullptr ||
                content_manager.get_current()->character_images != content_manager.get_previous()->character_images)
            {
                if (content_manager.get_current()->character_images.empty())
                    result_ss << TAG_STARTS << "hide" << TAG_ENDS << "*\n";
                else if (content_manager.get_previous() == nullptr ||
                         content_manager.get_previous()->character_images.empty())
                {
                    result_ss << TAG_STARTS << "display" << TAG_ENDS;
                    for (const std::string &character_name: content_manager.get_current()->character_images)
                        result_ss << Naming(character_name).get_name() << " ";
                    result_ss << "\n";
                }
            }

            if (content_manager.get_current()->has_next())
            {
                if (content_manager.get_current()->next.has_single_target())
                {
                    content_manager.set_id(content_manager.get_current()->next.get_target());
                } else
                {
                    // branch is currently not supported
                }
            } else
            {
                break;
            }
        }
    }

    // Write the stop symbol
    result_ss << "\n" << TAG_STARTS << "end" << TAG_ENDS << "\n";

    // Save the decompiled script
    std::ofstream out(out_path);
    out << result_ss.str();
    out.close();
}
